src/main/scala/calc/Compiler.scala

```diff
@@ -61,21 +62,113 @@ object Compiler {
-    // TODO
+  def compileExpr(tree: Tree, typeEnv: Map[String, irtpe.Type])(implicit mathModule: irt.Tree): irt.Tree = {
     implicit val pos = tree.pos
+    val functionTypes:Map[String, Int] = Map("sin" -> 1,
+      "cos" -> 1,
+      "tan" -> 1,
+      "log" -> 1,
+      "pow" -> 2,
+      "abs" -> 1,
+      "max" -> 2,
+      "min" -> 2)
 
     tree match {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
-      case _ =>
-        throw new Exception(
+      case BinaryOp(op, lhs, rhs) =>
+        val jsop = op match {
+          case "+" => irt.BinaryOp.Double_+
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, compileExpr(lhs, typeEnv), compileExpr(rhs, typeEnv))
+
+      case Let(ident, value, body) =>
+        val jsV = compileExpr(value, typeEnv)
+        val jsIdent = irt.Ident(ident.name)
+        val jsId = irt.VarDef(jsIdent, jsV.tpe, mutable = false, jsV)
+        irt.Block(List(jsId, compileExpr(body, typeEnv + (ident.name -> jsV.tpe))))
+
+      case Ident(n) =>
+        try {
+          irt.VarRef(irt.Ident(n))(typeEnv(n))
+        }catch{
+          case e:Exception => throw new Exception("No such identitier!")
+        }
+
+      case If(cond, thenp, elsep) =>
+        val jsThenp = compileExpr(thenp, typeEnv)
+        val jsElsep = compileExpr(elsep, typeEnv)
+        val jsCond = compileExpr(cond, typeEnv)
+        val finalCond = irt.BinaryOp(irt.BinaryOp.!==, jsCond, irt.DoubleLiteral(0.0))
+        if (jsElsep.tpe != jsElsep.tpe){
+          throw new Exception(s"The then and also phase in the conditional statement must bse same type")
+        }
+        irt.If(finalCond, jsThenp, jsElsep)(jsThenp.tpe)
+
+      case Closure(idList, body) =>
+        val captureList = findCapture(tree).toList
+        val captureParam = captureList.map(s =>
+          irt.ParamDef(irt.Ident(s), typeEnv(s), mutable = false, rest = false))
+        val params = idList.map(s =>
+          irt.ParamDef(irt.Ident("$$"+s.name), irtpe.AnyType, mutable = false, rest = false))
+        val newTypeEnv = idList.foldLeft(typeEnv)((env: Map[String, irtpe.Type], s:Ident)
+                                                    => env+(s.name -> irtpe.DoubleType))
+        val unboxParams = idList.map(s =>
+          irt.VarDef(irt.Ident(s.name), irtpe.DoubleType, mutable = false,
+            irt.Unbox(irt.VarRef(irt.Ident("$$"+s.name))(irtpe.AnyType), 'D')))
+        val bl = irt.Block(unboxParams ++ List(compileExpr(body, newTypeEnv)))
+        irt.Closure(captureParam, params, bl, captureList.map(s => irt.VarRef(irt.Ident(s))(typeEnv(s))))
+
+      case Call(f, args) =>
+        try {
+          val jsFunction = compileExpr(f, typeEnv)
+          irt.Unbox(irt.JSFunctionApply(jsFunction,args.map(s => compileExpr(s, typeEnv))), 'D')
+        }catch{
+          case e:Exception=> {
```

Comment:
This is a code smell. You're dealing with a completely normal case by throwing an exception and catching it. Worse, you're catching *any* type of `Exception`.

It would be much nicer to preemptively *test* whether `f` is a math function (and not a function in the lexical scope; which potentially shadows a standard function).

To do this, you might have to special-case a `case Call(Ident(name), args) if !typeEnv.contains(name) =>` or something like that.

This of course highlights the fact that you can only deal with direct calls to math functions (and not, e.g., `let f = sin in f(1)`), but that's ok. If you want an additional challenge, you could try to address that, but that's not necessary.